### PR TITLE
Save and load TensorMap from in-memory buffers

### LINKED ISSUE #94: Implement Python's pickle protocol for TensorMap
This would give us a better integration with the overall Python ecosystem. 

This should be done using `equistore.io.load`/`equistore.io.save`, but the functions themselves might require a bit of rework to write to (resp. load from) in-memory buffer instead of writing to (resp. loading from) the filesystem.

### PR DESCRIPTION
**Pull request by Luthaf**
Part of #133 
Fix #94

This allow implementing the Pickle protocol without writing to a file first, and to reduce the number of copies when loading data/sharing data with other python processes.

----
:books: Documentation preview :books:: https://equistore--298.org.readthedocs.build/en/298/

### CODE REVIEW

File: equistore-torch/src/misc.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include <equistore.hpp>
+#include <torch/types.h>
+
+#include "equistore/torch/array.hpp"
+#include "equistore/torch/misc.hpp"
+
+using namespace equistore_torch;
+
+
+eqs_status_t equistore_torch::details::create_torch_array(
```
Comment (Luthaf) [inline equistore-torch/src/misc.cpp:16]: Note for myself: check that we enclose this function in a try-catch block in the C++ API, or do it here

**Review by PicoCentauri — COMMENTED** [review]
I took a look. The c-part is a bit hard to follows for me but it looks reasonable.

File: equistore-torch/src/misc.cpp (hunk 1 above)
Comment (Luthaf) [reply equistore-torch/src/misc.cpp:12]: Done

File: equistore-torch/src/register.cpp [hunk 2]
```diff
@@ -180,15 +181,38 @@ TORCH_LIBRARY(equistore, m) {
...
-        //     },
-        //     // __setstate__
-        //     [](std::string state) -> torch::intrusive_ptr<TorchCalculator> {
-        //         // TODO
-        //     })
+        .def_pickle(
+            // __getstate__
+            [](const TorchTensorMap& self) -> torch::Tensor {
+                auto buffer = equistore::TensorMap::save_buffer(self->as_equistore());
+                // move the buffer to the heap so it can escape this function
```
Comment (agoscinski) [inline equistore-torch/src/register.cpp:188]: can we explain this a bit more? It was nontrivial for me that from_blob does not take ownership, that is why we need to put on the heap.

```suggestion
                // move the buffer to the heap so it can escape this function
                // since from_blob will not take ownership over the data
```

**Review by agoscinski — COMMENTED** [review]
Looks good. Just small comment

File: equistore-torch/src/register.cpp (hunk 2 above)
Comment (Luthaf) [reply equistore-torch/src/register.cpp:188]: done!
